### PR TITLE
fix(sequelize/query): retry entire query

### LINKED ISSUE #2113: Preferred way to handle reconnect
In case the db connection is lost, we need to reconnect to the database. I couldn't find a preferred way to handle this issue. Do you have a best practice? The issue can be solved on various ways:
- handle in sequelize
- handle in db module like pg
- handle this issue outside of nodejs with connection handler

Thanks in advance

### LINKED ISSUE #8014: ResourceRequest timed out - connection not reestablished after db restarted
## What you are doing?
We're running into an issue where Sequelize will not reconnect to the database after the database is stopped and started again.  If you want, you can try hosting a local mssql database using this tutorial and running the following code, and following the steps I provide below:  

https://medium.com/@reverentgeek/sql-server-running-on-a-mac-3efafda48861

```js
const options = {
  host: 'localhost',
  dialect: 'mssql',
  pool: {
    max: 1,
    min: 0,
    idle: 5000,
    acquire: 5000
  },
  dialectOptions: {
    requestTimeout: 5000
  },
  isolationLevel: 'READ UNCOMMITTED'
};

const Sequelize = require('sequelize')

const sequelize = new Sequelize('testing', 'sa', 'P@55w0rd', options);

setInterval(() => {
  sequelize
    .query("SELECT 'connection is successful'", {
      type: Sequelize.QueryTypes.SELECT
    })
    .then((results) => {
      console.log(results);
    })
    .catch((err) => {
      console.log(err);
    })
}, 5000);
```

Steps after code is running and mssql container is running:

1. stop the container
2. wait until an interval fires and throws an error
3. start the container
4. notice that all future queries result in timeout errors

I'm pretty sure this should reject, but I'm not too sure: https://github.com/sequelize/sequelize/blob/master/lib/dialects/abstract/connection-manager.js#L98

## What do you expect to happen?
Sequelize to reconnect and have queries work as normal.

## What is actually happening?
Sequelize fails to reconnect the created pool resource.

__Dialect:__ mssql, but probably all
__Database version:__ XXX
__Sequelize version:__ 4.4.2

### PR DESCRIPTION
**Pull request by sushantdhiman**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](../CONTRIBUTING.md)?

### Description of change

> 1. Wraps the entire query process in retry, instead of just the query.
> 2. Failed queries will yield the connection back to the pool, and request another one on the next retry
> This allows the retry functionality to work with the pool reconnect functionality :100: 

Continued from https://github.com/sequelize/sequelize/pull/8961

Closes #2113 
Closes #8022 
Closes #8014

### CODE REVIEW

File: test/unit/query.test.js [hunk 1]
```diff
@@ -1,37 +0,0 @@
-'use strict';
```
Comment (sushantdhiman) [inline test/unit/query.test.js]: As noted in original PR this test was added to ensure connection is only released once and there is no `Resource is not a part of pool` error.

Now each retry fetch new connection so this test is not required. Also that error was already handled [here](https://github.com/sequelize/sequelize/blob/d65a95c953ccd0595d5b406cd198ac149e14d955/lib/dialects/abstract/connection-manager.js#L287)

File: lib/sequelize.js [hunk 2]
```diff
@@ -517,41 +525,47 @@ class Sequelize {
...
 
       if (options.transaction && options.transaction.finished) {
-        const error = new Error(options.transaction.finished+' has been called on this transaction('+options.transaction.id+'), you can no longer use it. (The rejected query is attached as the \'sql\' property of this error)');
+        const error = new Error(`${options.transaction.finished} has been called on this transaction(${options.transaction.id}), you can no longer use it. (The rejected query is attached as the \'sql\' property of this error)`);
         error.sql = sql;
         return Promise.reject(error);
       }
 
-      if (this.test._trackRunningQueries) {
+      if (isFirstTry && this.test._trackRunningQueries) {
```
Comment (sushantdhiman) [inline lib/sequelize.js]: Not related to this PR but a possible fix for randomly failing tests due to tracking queries //cc @janmeier

**Comment by contra** [comment]
found one broken case here - retry-as-promised seems to start the timeout counter before any retries have happened, so if you try to run a really long query (lets say 120s) and you set the retry timeout for 60s it will fail with a timeout, even though 0 retries happened. seems like something that needs to be fixed in retry-as-promised.

**Comment by sushantdhiman** [comment]
> found one broken case here - retry-as-promised seems to start the timeout counter before any retries have happened, so if you try to run a really long query (lets say 120s) and you set the retry timeout for 60s it will fail with a timeout, even though 0 retries happened. seems like something that needs to be fixed in retry-as-promised.

Current default timeout is `undefined` which means no timeout. But with timeout if there is such bug it should be fixed in https://github.com/mickhansen/retry-as-promised

**Comment by contra** [comment]
@sushantdhiman Yep, understood where the fault lies - I think this can be merged without the fix, just wanted to mention it. I've opened the ticket there and will probably send a PR if I have time to https://github.com/mickhansen/retry-as-promised/issues/14.

File: lib/sequelize.js (hunk 2 above)
Comment (janmeier) [reply lib/sequelize.js]: Could very well be!

Comment (eseliger) [reply lib/sequelize.js]: that would be awesome!

**Comment by cladjules** [comment]
Great that is resolved :)
Do you know what version it's going to be released?

Thanks :)

**Comment by reiz** [comment]
We need this fix ASAP :-) We are affected by this and looking forward to this fix!

**Comment by eseliger** [comment]
https://github.com/sequelize/sequelize/releases/tag/v4.32.7 enjoy!

**Pull request merged**